sdk/java/pulumi/src/main/java/com/pulumi/core/Urn.java

```diff
@@ -1,75 +1,312 @@
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        Urn urn = (Urn) o;
+        return stack.equals(urn.stack) && project.equals(urn.project) && qualifiedType.equals(urn.qualifiedType) && name.equals(urn.name);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(stack, project, qualifiedType, name);
+    }
 
-    private Urn() {
-        throw new UnsupportedOperationException("static class");
+    @Override
+    public String toString() {
+        return new StringJoiner(", ", Urn.class.getSimpleName() + "[", "]")
+                .add("stack='" + stack + "'")
+                .add("project='" + project + "'")
+                .add("qualifiedType=" + qualifiedType)
+                .add("name='" + name + "'")
+                .toString();
+    }
+
+    @InternalUse
+    @SuppressWarnings("ConstantConditions") // IntelliJ can't understand the custom validators
+    public static Urn parse(String urn) {
+        require(u -> isNonEmptyOrNull(u), urn,
+                () -> format("expected urn to be not empty and not null, got: '%s'", urn)
+        );
+        require(u -> u.startsWith(Prefix), urn,
+                () -> format("expected urn to start with '%s', got: '%s'", Prefix, urn)
+        );
+
+        var urnParts = urn.split(PartsSeparator);
+        require(up -> up.length == 4, urnParts,
+                () -> format("expected urn to have 4 parts, separated by '%s', got '%s' in '%s'",
+                        PartsSeparator, urnParts.length, urn
+                )
+        );
+        var stack = urnParts[0].replaceFirst(Prefix, "");
+        require(s -> isNonEmptyOrNull(s), stack,
+                () -> format("expected urn stack part to be not empty, got: '%s'", stack)
+        );
+        require(s -> !s.contains(PartsSeparator), stack,
+                () -> format("expected urn stack part to not contain '%s', got: '%s'", ParentSeparator, stack)
+        );
+        var project = urnParts[1];
+        require(p -> isNonEmptyOrNull(p), project,
+                () -> format("expected urn project part to be not empty, got: '%s'", project)
+        );
+        require(p -> !p.contains(PartsSeparator), project,
+                () -> format("expected urn project part to not contain '%s', got: '%s'", ParentSeparator, project)
+        );
+        var qualifiedType = urnParts[2];
+        require(q -> isNonEmptyOrNull(q), qualifiedType,
+                () -> format("expected urn qualifiedType part to be not empty, got: '%s'", qualifiedType)
+        );
+        var name = urnParts[3];
+        require(n -> isNonEmptyOrNull(n), name,
+                () -> format("expected urn name part to be not empty, got: '%s'", name)
+        );
+
+        return new Urn(stack, project, QualifiedTypeName.parse(qualifiedType), name);
+    }
+
+    public String asString() {
+        return Prefix + this.stack + PartsSeparator
```

Comment:
Nit: StringJoiner here.

Reply:
but there's no join character, and after converting to string builder intellij suggests to convert back ;)

Reply:
Ah. OK then. I don't know if java optimizes long string + but in .NET that was a bit of a pain so I have a habit to avoid it. The strings are short here though.

Reply:
For the ultimate throughput StringBuilder will work better, but we don't have a hot loops or performance issues.
Probably safe to ignore the differences and go for readability:
https://stackoverflow.com/a/5188182/539481